Make pybind targets more modular, and combine C++ and Python pybind11 portions.

EricCousineau-TRI opened the pull request:
Relates #6760, implements [Step 2](https://github.com/RobotLocomotion/drake/issues/6760#issuecomment-337797746).

~This will be WIP until #7296 merges.~

\cc @sammy-tri 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7514)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@jwnimmer-tri for feature and platform review, please.

---

Review status: 0 of 40 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-KzyzbOD6S6qW6eZ1hNL:b-5a5hi0)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 45 of 73 files at r1, 22 of 28 files at r2, 6 of 8 files at r3.
Review status: 36 of 40 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/pydrake/__init__.py, line 9 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzz1AAjgt8M62N0PsaC:-Kzz1AAkcQJPE7biML-i:bp9jk8x) ([raw file](https://github.com/robotlocomotion/drake/blob/9dc22ee593292aa85c76ea8271ec8d8720978e81/drake/bindings/pydrake/__init__.py#L9)):*
> ```Python
> from . import common
> 
> from . import rbtree
> ```

I can understand why we want this to go away, but this has the potential to disturb downstream users.  Is there any way we could preemptively issue a deprecation warning here first, and then remove it in a month or two?  (I can't think of any way.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
(Posting on GitHub since Reviewable is down)

Looks like we can just override `getattr` for the module for the deprecation period, and then remove the shim once it's fully deprecated:
https://stackoverflow.com/a/7668273/7829525
This can also be used to import the module on-demand, still decoupling the initialization bits in `pydrake/__init__.py` from explicitly importing `rbtree`, keeping the dependencies streamlined.

jwnimmer-tri commented:
*[drake/bindings/pydrake/pybind.bzl, line 4 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzEasjIiX3bBzyKpOc:-KzzEasjIiX3bBzyKpOd:b-sqnyz6) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/pybind.bzl#L4)):*
> ```Python
> 
> # For some reason, `install` *must* be loaded via `@drake//` rather than
> # just `//`.
> ```

FYI This is because `InstallInfo = provider()` creates a singleton that we rely on being single, but loading with and without `@drake` creates two copies of it.

See bazelbuild/bazel#3493.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 73 files at r1, 2 of 8 files at r3, 3 of 7 files at r4, 1 of 1 files at r5.
Review status: 39 of 43 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/bindings/pydrake/BUILD.bazel, line 141 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzGx7SIDawYU0FgkzO:-KzzGx7SIDawYU0FgkzP:b-7oeplr) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/BUILD.bazel#L141)):*
> ```
>         ":symbolic_py",
>         "//drake/bindings/pydrake/solvers",
>     ],
> ```

This seems to duplicate the (inner list of the) deps in the above rule.  Perhaps a named constant?

---

*[drake/bindings/pydrake/BUILD.bazel, line 58 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzndwCuqbpPUTpAURJ:-KzzndwCuqbpPUTpAURK:bbczt6r) ([raw file](https://github.com/robotlocomotion/drake/blob/7eea5714eb0f08f52185435fc71a0032aa327ce6/drake/bindings/pydrake/BUILD.bazel#L58)):*
> ```
>     # deps = [
>     #     "//drake/common:autodiff",
>     # ],
> ```

Don't commit commented-out code.  At worst, you could just list the notional label inline as part of the comment prose.

---

*[drake/bindings/pydrake/path.py, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzHv4Ua8s1WdPHnEVH:-KzzHv4Ua8s1WdPHnEVI:b-aav5f5) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/path.py#L6)):*
> ```Python
> 
> def getDrakePath():
>     return pydrake.common.GetDrakePath()
> ```

At this point, `path.py` seems to be serving no purpose -- I think its only conceivable purpose would have been keeping a consistent filename versus when the implementation here used to be codegen'd, but this PR undoes even that putative benefit.  It seems like the aliasing here should just be rolled into its one point-of-use, and this file removed?

Or if there is a purpose I'm missing, maybe add a comment?

---

*[drake/bindings/pydrake/pybind.bzl, line 77 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzJByajAgYmfGdcKsM:-KzzJByajAgYmfGdcKsN:bkluomn) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/pybind.bzl#L77)):*
> ```Python
>         statically-linked libraries.
>     @param cc_devel_deps (optional)
>         C++ development dependencies.
> ```

So this attribute is currently unused, and is documented just as "development dependencies", which is ambiguous enough that nobody is going to be able to populate it correctly.  Would it be simpler to just remove it from this PR, instead of trying to repair it here?  (Documenting it; failing-fast when misused; etc.)

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 94 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzzp9CwgimgqYi12Brz:-Kzzp9CwgimgqYi12Bs-:b-keypef) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L94)):*
> ```
>         ":mathematicalprogram_py",
>         ":mosek_py",
>     ],
> ```

Unclear why `":module_py",` is omitted here.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 108 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzpR8Io4hzqPvyiuJI:-KzzpR8Io4hzqPvyiuJJ:b1okbfx) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L108)):*
> ```
>         ":gurobi_py",
>         ":ipopt_py",
>         ":mosek_py",
> ```

Alpha-sort?

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 108 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzpRqZosHAyw7bqJ3j:-KzzpRqZosHAyw7bqJ3k:b-7oeplr) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L108)):*
> ```
>         ":gurobi_py",
>         ":ipopt_py",
>         ":mosek_py",
> ```

This seems to duplicate the (inner list of the) deps in the above rule.  Perhaps a named constant?

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 113 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzporCZ9Vj6SjuyLN_:-KzzporCZ9Vj6SjuyLNa:byd8os3) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L113)):*
> ```
> 
> drake_py_test(
>     name = "mathematical_program_gurobi_solver_test",
> ```

_Meta_ On the one hand, I see that `mathematical_program_` prefix comes from the original BUILD rules prior to moving them into this file.  On the other hand, I see no reason to keep that as part of the name given everything else that is changing.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 143 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzqFlI_ff-nEn_oVOm:-KzzqFlI_ff-nEn_oVOn:bdlf5rd) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L143)):*
> ```
> 
> drake_py_test(
>     name = "mathematical_program_test",
> ```

FYI We might as well make the label match the filename (no first underscore)?

---

*[drake/bindings/pydrake/test/test_urdf_parser.py, line 5 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzzmx21mc62ih9mI4ZE:-Kzzmx21mc62ih9mI4ZF:b-wgi94w) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/test/test_urdf_parser.py#L5)):*
> ```Python
> import unittest
> import pydrake
> import pydrake.rbtree
> ```

FYI I actually thought the import as of r3 here was an improvement -- in that our code should employ the proper spellings at all times (since people will copy their style in new code from here).

If you want a unit test for the old spelling, that seems like a separate dedicated test file would do that best?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 38 of 44 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/bindings/pydrake/__init__.py, line 9 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzz1AAjgt8M62N0PsaC:-Kzz2dj0JE5od0_2aWvW:b-f6mec9) ([raw file](https://github.com/robotlocomotion/drake/blob/9dc22ee593292aa85c76ea8271ec8d8720978e81/drake/bindings/pydrake/__init__.py#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I can understand why we want this to go away, but this has the potential to disturb downstream users.  Is there any way we could preemptively issue a deprecation warning here first, and then remove it in a month or two?  (I can't think of any way.)
</blockquote></details>

Done. Added the `ModuleShim` utility thing, some tests, and kept usage of `import pydrake; pydrake.rbtree` in `urdf_parsers_test`.

(This does not expose `rbtree` in `pydrake.__all__`. I can change this if need be; there is no way that I can think of that will warn about this usage, though.)

---

*[drake/bindings/pydrake/pybind.bzl, line 4 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzEasjIiX3bBzyKpOc:-KzznugS_9GCwyWkrnv5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/pybind.bzl#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This is because `InstallInfo = provider()` creates a singleton that we rely on being single, but loading with and without `@drake` creates two copies of it.

See bazelbuild/bazel#3493.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Addressed part of the reviews... Unfortunately, Reviewable did not like me squashing changes into intermediate commits, and I lost track of the revision-based diffs. I've just switched it to "Combine commits for review" for now.

---

Review status: 33 of 89 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/bindings/pydrake/BUILD.bazel, line 141 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzGx7SIDawYU0FgkzO:-Kzzx-r8MOv0jD5QqWRX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/BUILD.bazel#L141)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This seems to duplicate the (inner list of the) deps in the above rule.  Perhaps a named constant?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/BUILD.bazel, line 58 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzndwCuqbpPUTpAURJ:-KzzwyToPi1GwgaasPJi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7eea5714eb0f08f52185435fc71a0032aa327ce6/drake/bindings/pydrake/BUILD.bazel#L58)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Don't commit commented-out code.  At worst, you could just list the notional label inline as part of the comment prose.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 77 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzJByajAgYmfGdcKsM:-Kzzx6pPeFR6owjD7U6O:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/pybind.bzl#L77)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So this attribute is currently unused, and is documented just as "development dependencies", which is ambiguous enough that nobody is going to be able to populate it correctly.  Would it be simpler to just remove it from this PR, instead of trying to repair it here?  (Documenting it; failing-fast when misused; etc.)
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 94 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzzp9CwgimgqYi12Brz:-Kzzx9MTkzL_0Etzsdk_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L94)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why `":module_py",` is omitted here.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 108 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzpR8Io4hzqPvyiuJI:-KzzxA01a0U9-eB7Sisr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Alpha-sort?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 108 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzpRqZosHAyw7bqJ3j:-KzzxARKUPSJCEmVtNfL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This seems to duplicate the (inner list of the) deps in the above rule.  Perhaps a named constant?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 113 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzporCZ9Vj6SjuyLN_:-KzzxAsGhC3L4b2ccF03:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ On the one hand, I see that `mathematical_program_` prefix comes from the original BUILD rules prior to moving them into this file.  On the other hand, I see no reason to keep that as part of the name given everything else that is changing.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 143 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzqFlI_ff-nEn_oVOm:-KzzxBNoE94aizVVLBIi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/solvers/BUILD.bazel#L143)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI We might as well make the label match the filename (no first underscore)?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/test/test_urdf_parser.py, line 5 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-Kzzmx21mc62ih9mI4ZE:-KzzxC_2sn75Bt2BC8Kn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/test/test_urdf_parser.py#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I actually thought the import as of r3 here was an improvement -- in that our code should employ the proper spellings at all times (since people will copy their style in new code from here).

If you want a unit test for the old spelling, that seems like a separate dedicated test file would do that best?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/path.py, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzHv4Ua8s1WdPHnEVH:-Kzzyiz0_6nz52jurYOQ:b-74bepi) ([raw file](https://github.com/robotlocomotion/drake/blob/e04cc96fc7e56b82db052ac3e0f1f9ed69084b7f/drake/bindings/pydrake/path.py#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

At this point, `path.py` seems to be serving no purpose -- I think its only conceivable purpose would have been keeping a consistent filename versus when the implementation here used to be codegen'd, but this PR undoes even that putative benefit.  It seems like the aliasing here should just be rolled into its one point-of-use, and this file removed?

Or if there is a purpose I'm missing, maybe add a comment?
</blockquote></details>

Done. Removed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-KzzxJuBAsifbdi9TS92:bvns3yh)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
It's okay, I was reviewing as a monolith anyway.  I was going to peek at the curation once all the edits are finished.

---

Reviewed 1 of 73 files at r1, 1 of 5 files at r6, 4 of 53 files at r7, 4 of 31 files at r8.
Review status: 41 of 45 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--1vi_oPSKK233c8Yy:-L--1viafSwVJsvVrqbZ:b-ehjc48) (no related file):*
Reminder to myself to inspect and test the installed image.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--1yXw46Zn6OqdV63O:-L--1yXw46Zn6OqdV63P:b-eu0sfa) (no related file):*
Reminder to myself to double-check the commit curation.

---

*[drake/bindings/pydrake/pybind.bzl, line 25 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzrV3HPtQa94Qauf3K:-KzzrV3HPtQa94Qauf3L:b-og0zaw) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L25)):*
> ```Python
> 
> def _drake_pybind_cc_binary(name, srcs = [], copts = [],
>                             deps = [], visibility = None):
> ```

Convention is to do one-arg-per-line once we need multiple lines.  Also, check on e.g. drake_cc.bzl for how that line-wraps.

---

*[drake/bindings/pydrake/pybind.bzl, line 65 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzruaULEDqzNRMoi8n:-KzzruaULEDqzNRMoi8o:bd4rqa5) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L65)):*
> ```Python
>                          py_srcs = [], py_deps = [],
>                          py_imports = [], add_install = True,
>                          visibility = None):
> ```

Convention is to do one-arg-per-line once we need multiple lines.  Also, check on e.g. drake_cc.bzl for how that line-wraps.  (All of this should be must further left.)

---

*[drake/bindings/pydrake/pybind.bzl, line 66 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzuNlVeWBU_qcQenDZ:-KzzuNlVeWBU_qcQenD_:bk9s7p) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L66)):*
> ```Python
>                          py_imports = [], add_install = True,
>                          visibility = None):
>     """ Declare a pybind11 library, with C++ base code and Python interface
> ```

_Meta_ I just noticed that many docstrings here have a space between the triple-quote and the first letter.  That seems unlike any PEP257 example.

---

*[drake/bindings/pydrake/pybind.bzl, line 94 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzsXolbHIKl7Pd-T8d:-KzzsXolbHIKl7Pd-T8e:b-r1g40f) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L94)):*
> ```Python
>     if not cc_so_name:
>         cc_so_name = "_" + name
>     if not cc_so_name.endswith('.so'):
> ```

There's no reason to allow for two spellings for name (with or without so), is there?  It seems like that would lead to confusing inconsistency.  Might as well just pick one and have this either unconditionally add the suffix, or move the hard-coded suffix to call sites.

---

*[drake/bindings/pydrake/pybind.bzl, line 143 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzuqOtAcajSVjeT-fI:-KzzuqOtAcajSVjeT-fJ:bsuhzbp) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L143)):*
> ```Python
>         return target + ":install"
> 
> def get_pybind_pkg_dest(py_pkg_install = None):
> ```

_Meta_ `pkg` is not an allowed abbreviation per C++ GSG, which we've (I've?) been informally applying to python code as well.  (I think the same reasoning applies.)

---

*[drake/bindings/pydrake/pybind.bzl, line 164 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzztoiP9aUfQHoYa30H:-KzztoiP9aUfQHoYa30I:bxn4y8g) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L164)):*
> ```Python
>     # Make the number of parent directories.
>     rel_path = "/".join([".."] * num_pieces)
>     return (rel_path, sub_pkg)
> ```

FYI `return struct(rel_path=rel_path, sub_pkg=sub_pkg)` here would possibly be more readable at call sites.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 39 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--2TZsPWp7okFpN-oG:-L--2TZtagA7SQhNKIAW:b-7n5ovt) ([raw file](https://github.com/robotlocomotion/drake/blob/53a948ad59cae3bef9130cd428fc4575458028df/drake/bindings/pydrake/solvers/BUILD.bazel#L39)):*
> ```
> 
> # TODO(eric.cousineau): Rename `mathematicalprogram` to
> # `mathematical_program`.
> ```

FYI Maybe defer that until the `pydrake` -> `drake` rename?  Otherwise its just one more compat alias to keep around (probably?).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-L--1UB6NMK9pfpkEA9I:b-dwujxx)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 8 files at r4, 4 of 5 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/bindings/pydrake/test/module_shim_example/sub_module.py, line 0 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--3WLzxHd_VXi36hpD:-L--3WLzxHd_VXi36hpE:biqvkho) ([raw file](https://github.com/robotlocomotion/drake/blob/53a948ad59cae3bef9130cd428fc4575458028df/drake/bindings/pydrake/test/module_shim_example/sub_module.py#L0)):*
Files must not be empty.

http://drake.mit.edu/code_style_guide.html#additional-rules-

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I think there's enough here to merit a second reviewer.  We'll count me as feature review, so platform reviewer would be good once the next round of fixes are in.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-L--4V5GHaFoWitY4Ncr:b5nlk7p)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Checkpoint. (Laptop battery running low, plane power outlets insufficient for charging... Gonna be a long flight.)
Commits got messy. May try later to sanitize. If not, will squash.

+@sammy-tri for platform review, please.

(NOTE: I would like to rename `drake_pybind_library` to `drake_pybind_module`, to keep the name consistent with other portions; that or consistently use `library` instead of `module` in the Bazel code.)

---

Review status: 41 of 44 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/bindings/pydrake/pybind.bzl, line 25 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzrV3HPtQa94Qauf3K:-L-3G_0baslDxo4ot6rX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L25)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Convention is to do one-arg-per-line once we need multiple lines.  Also, check on e.g. drake_cc.bzl for how that line-wraps.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 65 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzruaULEDqzNRMoi8n:-L-3Ga7tHlhckUparE6o:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Convention is to do one-arg-per-line once we need multiple lines.  Also, check on e.g. drake_cc.bzl for how that line-wraps.  (All of this should be must further left.)
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 66 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzuNlVeWBU_qcQenDZ:-L-3Gb7IErBVxT5prVvT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ I just noticed that many docstrings here have a space between the triple-quote and the first letter.  That seems unlike any PEP257 example.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 94 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzsXolbHIKl7Pd-T8d:-L-3GcWgy4QbTo8QUoUI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L94)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

There's no reason to allow for two spellings for name (with or without so), is there?  It seems like that would lead to confusing inconsistency.  Might as well just pick one and have this either unconditionally add the suffix, or move the hard-coded suffix to call sites.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 143 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzzuqOtAcajSVjeT-fI:-L-3Gh3yA7OA0FLqzN6_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L143)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ `pkg` is not an allowed abbreviation per C++ GSG, which we've (I've?) been informally applying to python code as well.  (I think the same reasoning applies.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-L-3GmwbPJXXTdV5_goh:b-aujmdg)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 39 of 44 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/bindings/pydrake/BUILD.bazel, line 6 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L-3q-TgRH5HAXr8FiMA:-L-TspIndoRh89kZQjDN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e050e94994961d360450d07181f962476a35a458/drake/bindings/pydrake/BUILD.bazel#L6)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Unused after this change?
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/pybind.bzl, line 164 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-KzztoiP9aUfQHoYa30H:-L-TsrB1XJivaJ2RPLSV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29eca2d0d6531409511dabd54ad96ac4b8cac17a/drake/bindings/pydrake/pybind.bzl#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `return struct(rel_path=rel_path, sub_pkg=sub_pkg)` here would possibly be more readable at call sites.
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/solvers/BUILD.bazel, line 39 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--2TZsPWp7okFpN-oG:-L-TsrrpxkG9-ZOLOx0f:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/53a948ad59cae3bef9130cd428fc4575458028df/drake/bindings/pydrake/solvers/BUILD.bazel#L39)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Maybe defer that until the `pydrake` -> `drake` rename?  Otherwise its just one more compat alias to keep around (probably?).
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/test/module_shim_test.py, line 57 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L-3t3Y8JogPni-qNICN:-L-TssoK4ek8pOn6SWY2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e050e94994961d360450d07181f962476a35a458/drake/bindings/pydrake/test/module_shim_test.py#L57)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW consider a comment for why this isn't a function in `TestModuleShim`
</blockquote></details>

Done.

---

*[drake/bindings/pydrake/test/module_shim_example/sub_module.py, line  at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L--3WLzxHd_VXi36hpD:-L-TsuuU9hq8rCB1TsKZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/53a948ad59cae3bef9130cd428fc4575458028df/drake/bindings/pydrake/test/module_shim_example/sub_module.py#L)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Files must not be empty.

http://drake.mit.edu/code_style_guide.html#additional-rules-
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 39 of 44 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/bindings/pydrake/BUILD.bazel, line 4 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-L-3puFn1WG4bVAjwO08:-L-Tt2CWAaPxDqQp-Pl9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e050e94994961d360450d07181f962476a35a458/drake/bindings/pydrake/BUILD.bazel#L4)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Unused after this change?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7514#-:-L-n0oFtI0GVWyeocUrM:bnfp4nl)*
<!-- Sent from Reviewable.io -->
